File: sdk/python/bunqueue/flow_plan_legacy.py
"""Pure atomic planners for FlowProducer chains and fan-in graphs."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, List, Sequence, Set

from .flow_plan import (
    MAX_FLOW_JOBS,
    AtomicFlowJob,
    FlowIdFactory,
    allocate_flow_id,
    assert_atomic_options,
    assert_flow_identity,
    flow_data,
    flow_input,
    portable_flow_id,
)


@dataclass
class ChainPlan:
    jobs: List[AtomicFlowJob]
    ids: List[str]


@dataclass
class FanInPlan:
    jobs: List[AtomicFlowJob]
    parallel_ids: List[str]
    final_id: str


def _allocate_steps(
    steps: Sequence[Dict[str, Any]], id_factory: FlowIdFactory
) -> List[str]:
    if len(steps) > MAX_FLOW_JOBS:
        raise ValueError("flow exceeds the 10000 job limit")
    allocated: Set[str] = set()
    ids: List[str] = []
    for step in steps:
        assert_flow_identity(step)
        if step.get("children"):
            raise ValueError("nested children are not supported by this flow method")
        raw_opts = step.get("opts")
        opts = {} if raw_opts is None else raw_opts
        assert_atomic_options(opts)
        ids.append(allocate_flow_id(opts, allocated, id_factory))
    return ids


def plan_chain(
    steps: Sequence[Dict[str, Any]],
    id_factory: FlowIdFactory = portable_flow_id,
) -> ChainPlan:
    ids = _allocate_steps(steps, id_factory)
    jobs: List[AtomicFlowJob] = []
    for index, step in enumerate(steps):
        dependency = ids[index - 1] if index else None
        jobs.append(
            {
                "id": ids[index],
                "queue": step["queueName"],
                "input": flow_input(
                    flow_data(
                        step["name"],
                        step.get("data"),
                        {"__flowParentId": dependency},
                    ),
                    step.get("opts") or {},
                    depends_on=[dependency] if dependency else None,
                ),
            }
        )
    return ChainPlan(jobs, ids)


def plan_bulk_then(
    parallel: Sequence[Dict[str, Any]],
    final: Dict[str, Any],
    id_factory: FlowIdFactory = portable_flow_id,
) -> FanInPlan:
    steps = [*parallel, final]
    ids = _allocate_steps(steps, id_factory)
    parallel_ids = ids[:-1]
    final_id = ids[-1]
    parallel_jobs = [
        {
            "id": parallel_ids[index],
            "queue": step["queueName"],
            "input": flow_input(
                flow_data(
                    step["name"],
                    step.get("data"),
                    {
                        "__parentId": final_id,
                        "__parentQueue": final["queueName"],
                    },
                ),
                step.get("opts") or {},
                parent_id=final_id,
            ),
        }
        for index, step in enumerate(parallel)
    ]
    final_job = {
        "id": final_id,
        "queue": final["queueName"],
        "input": flow_input(
            flow_data(
                final["name"],
                final.get("data"),
                {
                    "__flowParentIds": parallel_ids,
                    "__childrenIds": parallel_ids,
                },
            ),
            final.get("opts") or {},
            depends_on=parallel_ids,
            children_ids=parallel_ids,
        ),
    }
    return FanInPlan([*parallel_jobs, final_job], parallel_ids, final_id)

File: sdk/typescript/src/flow-plan-legacy.ts
import { randomUUID } from 'node:crypto';
import {
  type AtomicFlowJobInput,
  allocateFlowId,
  type FlowIdFactory,
  flowData,
  flowInput,
} from './flow-plan.js';
import type { FlowStep } from './flow-types.js';

const MAX_FLOW_JOBS = 10_000;
const QUEUE_RE = /^[a-zA-Z0-9_\-.:]+$/;

export interface ChainPlan {
  jobs: AtomicFlowJobInput[];
  ids: string[];
}

export interface FanInPlan {
  jobs: AtomicFlowJobInput[];
  parallelIds: string[];
  finalId: string;
}

function allocateSteps<T>(
  steps: FlowStep<T>[],
  ids: Set<string>,
  idFactory: FlowIdFactory
): string[] {
  if (steps.length > MAX_FLOW_JOBS) {
    throw new Error(`flow exceeds the ${MAX_FLOW_JOBS} job limit`);
  }
  return steps.map((step) => {
    if (!step || typeof step !== 'object') throw new Error('flow step must be an object');
    if (typeof step.name !== 'string' || !step.name || step.name.length > 256) {
      throw new Error('flow job name must be a non-empty string of at most 256 characters');
    }
    if (
      typeof step.queueName !== 'string' ||
      !step.queueName ||
      step.queueName.length > 256 ||
      !QUEUE_RE.test(step.queueName)
    ) {
      throw new Error('flow queueName is invalid');
    }
    const children = (step as FlowStep<T> & { children?: unknown }).children;
    if (children !== undefined && !Array.isArray(children)) {
      throw new Error('flow children must be an array');
    }
    if ((children?.length ?? 0) > 0) {
      throw new Error('nested children are not supported by this flow method');
    }
    return allocateFlowId(step.opts ?? {}, ids, idFactory);
  });
}

export function planChain<T>(
  steps: FlowStep<T>[],
  idFactory: FlowIdFactory = randomUUID
): ChainPlan {
  const ids = allocateSteps(steps, new Set(), idFactory);
  const jobs = steps.map((step, index): AtomicFlowJobInput => {
    const dependency = index > 0 ? ids[index - 1] : undefined;
    const data = flowData(step.name, step.data, {
      __flowParentId: dependency ?? null,
    });
    return {
      id: ids[index],
      queue: step.queueName,
      input: flowInput(data, step.opts ?? {}, {
        dependsOn: dependency ? [dependency] : undefined,
      }),
    };
  });
  return { jobs, ids };
}

export function planBulkThen<T>(
  parallel: FlowStep<T>[],
  final: FlowStep<T>,
  idFactory: FlowIdFactory = randomUUID
): FanInPlan {
  if (parallel.length >= MAX_FLOW_JOBS) {
    throw new Error(`flow exceeds the ${MAX_FLOW_JOBS} job limit`);
  }
  const ids = new Set<string>();
  const parallelIds = allocateSteps(parallel, ids, idFactory);
  const [finalId] = allocateSteps([final], ids, idFactory);
  const parallelJobs = parallel.map((step, index): AtomicFlowJobInput => ({
    id: parallelIds[index],
    queue: step.queueName,
    input: flowInput(
      flowData(step.name, step.data, {
        __parentId: finalId,
        __parentQueue: final.queueName,
      }),
      step.opts ?? {},
      { parentId: finalId }
    ),
  }));
  const finalJob: AtomicFlowJobInput = {
    id: finalId,
    queue: final.queueName,
    input: flowInput(
      flowData(final.name, final.data, {
        __flowParentIds: parallelIds,
        __childrenIds: parallelIds,
      }),
      final.opts ?? {},
      {
        dependsOn: parallelIds.length > 0 ? parallelIds : undefined,
        childrenIds: parallelIds.length > 0 ? parallelIds : undefined,
      }
    ),
  };
  return { jobs: [...parallelJobs, finalJob], parallelIds, finalId };
}
